### ISSUE TITLE
ARM Linux

### ISSUE THREAD

**Issue by xcracing** [MarcFletcher/NetworkComms.Net#45]
With the usefulness of RaspberryPi and clones/offshoots as well as the cross platform nature of .net core these days it would be super cool if NetworkComms.net could be used on Linux.

Use case - I developed a UWP program to run on Raspberry Pi that initially ran on Windows 10 IOT and connected to a host program on a windows based computer.   It uses NetworkComms.net for the host program to communicate with the RPi.  Has been working great.

Since Win10 IOT was never really further developed, seems mostly abandoned, and doesn't support the plethora of hardware that Linux does, I decided to see about porting my RPi based program to Linux and .net core 7.   Things were going great until I tried to open my UDP broadcast listener on the RPi.    I was met with an error message:

Socket.IOControl handles Windows-specific control codes and is not supported on this platform.

In googling around a bit it seems like this is maybe related to a keepalive setting that can be set another way.   In looking at the NetworkComms.net source, I only see one call to IOControl - and it's not applied to Android & iOS because the behavior is not required - could Linux be added to the clause to solve the issue for the same reason or am I running up against something else like the keepalive issue mentioned above?  Below is the code snippet (code tag doesn't seem to be working properly so I'll just quote)

> #if !ANDROID && !iOS
            //By default we ignore ICMP destination unreachable packets so that we can continue to use the udp client even if we send something down a black hole
            //This is unsupported in Mono but also not required as the same behaviour is not observed.
            if (UDPConnection.IgnoreICMPDestinationUnreachable && Type.GetType("Mono.Runtime") == null)
                this.udpClient.Client.IOControl(SIO_UDP_CONNRESET, new byte[] { 0, 0, 0, 0 }, new byte[] { 0, 0, 0, 0 });
#endif

**Comment by xcracing** [comment]
I downloaded the source, put the files in a project set for .netstandard 2.0 and deleted the lines with IOControl just to see if that would do the trick.   My application runs without errors now, but doesn't trigger my packetHandler event when UDP packets are received.   I'm assuming there is other code that needs to be changed to work with .net standard 2.0.   Is there source somewhere for the 3.0.4.1 .net standard 2.0 NuGet package - or any other thoughts?

**Comment by xcracing** [comment]
Hi Mark, thanks for sharing the great library and the prompt response. 

I have a known good .net 4.0 application using NetworkComms to send UDP broadcasts out.   I also have know good receiver apps on Win10IOT/UWP and Windows .net 4.8/Winforms that are currently working (everything on the same broadcast domain, of course).  These apps are more complex than what I'm currently trying to get working as they use the UDP broadcasts to discover the sender then setup TCP connections to communicate.   

I'm trying to get my Linux port of my Win10IOT app to receive the broadcasts and spit out a console message that it sees the broadcast that occurs every 5 sec - which should be very simple based on my experience and existing code using NetworkComms.   Basically, it's a very simple proof that the library will work on this platform.   I'm logging all the code and everything seems to setup as it should & error free, it's just that once I setup the connection, nothing else happens.  I've also built a separate Linux app using .net's built in UDPClient to receive the broadcasts - and it does see them.

I've used known good code to setup my packet handlers in my Linux app.   I've also used a command in Linux to see the port open on my application and I can use a command to actually see the broadcast packets arriving at the Linux device being tested in realtime, so I'm nearly certain the issue is with the library I've compiled and am attempting to use on Linux.

.net 7:  My Linux app is .net 7, but it can use .netstandard 2.0 libraries, so as long as I get get NetworkComms.net working on .net standard 2.0 and not using any unsupported functions in Linux, I believe it should work.   Unfortunately, I can't currently use Mono without scrapping a bunch of work and taking a different route on the UI.

I tried the NuGet .netstandard 2.0 build of NetworkComms (v3.0.4.1) and that's what gives me the IOControl error.   All I could find on Github was the .net 4 source for NetworkComms - so I attempted to create a .net standard 2.0 project and brought all the code over from Github.   I commented the IOControl line and had to remove some code regarding AppDomain so it would compile - but I'm not sure if other changes have to be made to get the .net Standard 2.0 version working as I was unable to find the source that the v3.0.4.1 version on NuGet was built from.

I turned on logging and here's what I'm seeing from the Linux app.   Note that I'm firing up several packetHandlers but am only testing the "beacon" currently.  I also added an unmanaged handler, just in case.  Note the port I'm opening is 48198

> 1.21.40.229 [1 - Info] - Added incoming packetHandler for 'beacon' packetType.
1.21.40.229 [1 - Info] - Added incoming packetHandler for 'jsonMsg' packetType.
1.21.40.230 [1 - Info] - Added incoming packetHandler for 'clrMsg' packetType.
1.21.40.231 [1 - Info] - Added globalConnectionShutdownDelegates. 1
1.21.40.235 [1 - Info] - Added incoming packetHandler for 'Unmanaged' packetType.
1.21.40.236 [1 - Info] - Added globalConnectionEstablishDelegates. 1
1.21.40.291 [1 - Info] - Created new connection listener (UDP-E).
1.21.40.292 [1 - Info] - Created new connection listener (UDP-E).
1.21.40.293 [1 - Info] - Created new connection listener (UDP-E).
1.21.40.339 [1 - Info] - Appended connection specific packet handlers from listener '[UDP-E] Not Listening' to connection '[UDP-E-U] 127.0.0.1:48198 -> 0.0.0.0:0'.
1.21.40.342 [1 - Trace] - Checking for existing connection by remoteEndPoint='0.0.0.0:0', localEndPoint='127.0.0.1:48198', connectionType='UDP' and ApplicationLayerProtocolStatus='Enabled'.
1.21.40.363 [1 - Trace] - RetrieveConnection by remoteEndPoint='0.0.0.0:0', localEndPoint='127.0.0.1:48198', connectionType='UDP' and ApplicationLayerProtocolStatus='Enabled'. No matching connections found.
1.21.40.382 [1 - Trace] - Adding connection reference by endPoint. Connection='[UDP-E-U] 127.0.0.1:48198 -> 0.0.0.0:0'.
1.21.40.383 [1 - Trace] - Creating new UDPConnection with [UDP-E-U] 127.0.0.1:48198 -> 0.0.0.0:0
1.21.40.439 [1 - Info] - Listener started (UDP-E - 127.0.0.1:48198).
1.21.40.450 [1 - Info] - Appended connection specific packet handlers from listener '[UDP-E] Not Listening' to connection '[UDP-E-U] [::1]:48198 -> 0.0.0.0:0'.
1.21.40.450 [1 - Trace] - Checking for existing connection by remoteEndPoint='0.0.0.0:0', localEndPoint='[::1]:48198', connectionType='UDP' and ApplicationLayerProtocolStatus='Enabled'.
1.21.40.450 [1 - Trace] - RetrieveConnection by remoteEndPoint='0.0.0.0:0', localEndPoint='[::1]:48198', connectionType='UDP' and ApplicationLayerProtocolStatus='Enabled'. No matching connections found.
1.21.40.450 [1 - Trace] - Adding connection reference by endPoint. Connection='[UDP-E-U] [::1]:48198 -> 0.0.0.0:0'.
1.21.40.450 [1 - Trace] - Creating new UDPConnection with [UDP-E-U] [::1]:48198 -> 0.0.0.0:0
1.21.40.452 [1 - Info] - Listener started (UDP-E - [::1]:48198).
1.21.40.452 [1 - Info] - Appended connection specific packet handlers from listener '[UDP-E] Not Listening' to connection '[UDP-E-U] 192.168.xxx.160:48198 -> 0.0.0.0:0'.
1.21.40.452 [1 - Trace] - Checking for existing connection by remoteEndPoint='0.0.0.0:0', localEndPoint='192.168.xxx.160:48198', connectionType='UDP' and ApplicationLayerProtocolStatus='Enabled'.
1.21.40.452 [1 - Trace] - RetrieveConnection by remoteEndPoint='0.0.0.0:0', localEndPoint='192.168.xxx.160:48198', connectionType='UDP' and ApplicationLayerProtocolStatus='Enabled'. No matching connections found.
1.21.40.452 [1 - Trace] - Adding connection reference by endPoint. Connection='[UDP-E-U] 192.168.xxx.160:48198 -> 0.0.0.0:0'.
1.21.40.452 [1 - Trace] - Creating new UDPConnection with [UDP-E-U] 192.168.xxx.160:48198 -> 0.0.0.0:0
1.21.40.453 [1 - Info] - Listener started (UDP-E - 192.168.xxx.160:48198).

**Comment by xcracing** [comment]
So hoping you might have an idea for somewhere to look.

I compiled networkcomms to .net standard 2.0 and commented the offending code mentioned above that was stopping Linux from working.  It doesn't seem to be necessary on Linux (In fact, there's a comment there it's not required for Mono, which I assume is because of the underlying OS)

I built my .net 7 app that will run on Windows or ARM Linux & uses NetworkComms to communicate.   I also built a VERY quick test app to test raw System.Net.Sockets to on Linux - just to make sure the .net base library itself works as expected on Linux.

First I ran my test app on my Raspberry Pi - all it does is uses raw System.Sockets.Net to open a UDP port and listen for my server program's broadcasts using BeginReceive(), gets an AsyncCallBack when data is received, and outputs it on screen ---  Worked completely as expected.

Next, I built my real app which references NetworkCommsDotNet.    I ran the app on my test windows box first.  It detected the packets coming from the server and actually, the full protocol setup and worked completely as expected (i.e. it established the TCP communications channel and started talking).

I then ran the same exact code on the Raspberry Pi.  Everything sets up (i.e. listeners), no errors, I put debug logs in all the way to where BeginReceive() is called  and in the callback when packets are received, in NetworkComms - I see BeginReceive get called after the listener sets up, however, the AsyncCallback (IncomingUDPPacketHandler) is never called when the packets arrive.    I'm at a loss.   In pouring over NetworkComms source, at a very basic level, it's doing the exact same thing my simple test app is doing on the same exact device - setting up a UDP connection with BeginReceive and firing the async callback when data arrives.  But with NetworkComms the callback never gets called for some unknown reason....   I can't see anything wrong.  Hoping you might have an idea where to look or test next.

**Comment by xcracing** [comment]
Figured it out after many hours banging my head against the wall the last few days.   Maybe someone will have a better suggestion for a fix - especially a multi-platform friendly fix since this possibly breaks other platforms.

Turns out, when you listen with IPAddress.Any, Networkcomms.net iterates through the adapters and sets up listeners instead of using 0.0.0.0 (there's a note in the comments that says using Ipaddress.Any isn't supported on all platforms) --- for whatever reason, Arm Linux (and maybe other *nix?) appears to ignore broadcasts unless you use actual 0.0.0.0

One problem solved, I'm receiving my UDP broadcasts now.

Next issue, when you use IPAddress.Any - when you receive data, your local endpoint comes back as 0.0.0.0 and there's apparently no property or method to get the actual IP received.  My thought was to build something to try to figure it out from the sender's address - but turns out Marc did it for me by providing the IPTools.BestLocalEndPoint method (Thanks for that!).

So - in the end, I removed the conditional statement and loop in ConnectionListeners.cs that checks for IPAddress.Any and loops through available adapters to setup listeners on each.   Then in my code, I just get the actual LocalEndpoint using IPTools.BestLocalEndPoint and now my networking subsystem functions on ARM Linux running .net 7!!!!
